### PR TITLE
Implement LookAtModifier3D

### PR DESCRIPTION
**Pull request by TokageItLab**
Add a LookAtModifier3D. This modifier allows you to specify a forward axis and provides a two-axis Euler rotation to prevent roll the forward axis.

https://github.com/user-attachments/assets/33486536-142f-406a-b310-cb1d62c9c887

- Follow up https://github.com/godotengine/godot/pull/87888

# Features

## Time-based interpolation

When the character changes the lock-on target, it will take a certain amount of time to rotate instead of immediately rotating its head. The time-based interpolation uses a tween, so several transition/easing types can be selected.

https://github.com/user-attachments/assets/f0e18406-2be9-4695-92bf-aba8005cf7e6

## Angle limitation

This option allows you to prevent the character's neck from rotating 360 degrees. The limitation means that half angle of the limit from the bone rest is the range of motion. Both symmetrical and non-symmetrical limitations are supported.

Also, when this option is enabled, time-based interpolation will provide rest-aware rotation, similar to AnimationTree blending. This means that in some cases the shortest path will not be selected to prevent bones from penetrating through the body. See also [for-better-blending](https://docs.godotengine.org/en/stable/tutorials/animation/animation_tree.html#for-better-blending).

In addition, if the target moves outside of these limits, axis flipping may occur, where time-based interpolation is used.

https://github.com/user-attachments/assets/056b9503-cd11-4576-9148-7e160a84cab3

## Non-linear damping

To prevent abrupt changes in angular velocity at the edge of the angular limit, the option is to start non-linear(b-spline) damping from the threshold value.

This helps simulate natural human motion.

https://github.com/user-attachments/assets/10696033-0d39-42a7-b5f4-8a23872e90fa

## Origin setting

There are some cases where multiple bones need to always point in the same direction, such as the eyes.

For example, angle limitation may cause bone to flip outside their limits, which is not good because only one of the eyes will move significantly. Setting an arbitrary origin helps to solve this problem.

### Origin from

For example, To solve the above problem, it is possible to set the head to origin for eyes. Or you can use BoneAttachment or parent-child Node3D.

image

### Offset

The problem can be addressed by matching the origin by offset without any external node or bone as well.

https://github.com/user-attachments/assets/95d55acd-d140-415a-8287-f93e03ff5c10

### CODE REVIEW

**Comment by fire** [comment]
Does look at break bones with different body proportions?

**Comment by TokageItLab** [comment]
@fire You can set any axis as the forward axis. Note that forward axis does not mean the direction of the bone from parent to child, but rather the axis you want to target in the bone's rest.

For example, if retargeting is used, the recommended forward axis of the head is +Z in GodotHumanoid. Without retargeting, it depends on the rest in gltf.

As a related note to the PR in https://github.com/godotengine/godot/pull/97824, the forward axis of the GeneralSkeleton is +Z, but the forward axis of the OriginalSkeleton is -Z in the rest in the gltf and in the demo project in https://github.com/godotengine/godot/pull/97824#issuecomment-2394965894.

**Comment by fire** [comment]
How do I set parent to child?

File: scene/3d/look_at_modifier_3d.cpp [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+				origin_tr = bone_rest_space;
+			}
+		} else {
+			origin_tr = bone_rest_space;
+		}
+		forward_vector = bone_rest_space.basis.xform_inv((target->get_global_position() - origin_tr.translated_local(origin_offset).origin));
+		forward_vector_nrm = forward_vector.normalized();
+		if (forward_vector_nrm.abs().is_equal_approx(get_vector_from_axis(primary_rotation_axis))) {
+			destination = skeleton->get_bone_pose_rotation(bone);
+			forward_vector = Vector3(0, 0, 0); // Avoid to glitch.
```
Comment (fire) [inline scene/3d/look_at_modifier_3d.cpp:481]:
```suggestion
			forward_vector = Vector3(0, 0, 0); // Avoid animation glitch.
```

**Comment by TokageItLab** [comment]
Added:
- A configuration warning when the forward axis and rotation axis are parallel to the node
- Some comments to the code
- Some Notes to the document

**Review by SaracenOne — APPROVED** [review]
Discussed and tested during the Animation Meeting and everyone seems satisfied that it works and adds broadly useful functionality. 👍

It is also worth noting that it largely follows the spec put forward by the VRM consortium https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0/lookAt.md

File: doc/classes/LookAtModifier3D.xml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		</member>
+		<member name="origin_offset" type="Vector3" setter="set_origin_offset" getter="get_origin_offset" default="Vector3(0, 0, 0)">
+			The offset of the bone pose origin. Matching the origins by offset is useful for cases where multiple bones must always face the same direction, such as the eyes.
+			[b]Note:[/b] This value indicates the local position of the object set in [member origin_from].
+		</member>
+		<member name="origin_safe_margin" type="float" setter="set_origin_safe_margin" getter="get_origin_safe_margin" default="0.1">
+			If the target passes through too close to the origin than this value, time-based interpolation is used even if the target is within the angular limitations, to prevent the angular velocity from becoming too fast.
+		</member>
+		<member name="primary_damp_threshold" type="float" setter="set_primary_damp_threshold" getter="get_primary_damp_threshold">
+			The threshold to start damping for [member primary_limit_angle]. It provides non-linear(b-spline) interpolation, let it feel more resistance the more it rotate to the edge limit. This is useful for simulating the limits of human motion.
```
Comment (AThousandShips) [inline doc/classes/LookAtModifier3D.xml:67]:
```suggestion
			The threshold to start damping for [member primary_limit_angle]. It provides non-linear (b-spline) interpolation, let it feel more resistance the more it rotate to the edge limit. This is useful for simulating the limits of human motion.
```

File: scene/3d/look_at_modifier_3d.cpp [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+
+void LookAtModifier3D::set_bone(int p_bone) {
+	bone = p_bone;
+}
+
+int LookAtModifier3D::get_bone() const {
+	return bone;
+}
+
+void LookAtModifier3D::set_forward_axis(LookAtModifier3D::BoneAxis p_axis) {
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.cpp:86]:
```suggestion
void LookAtModifier3D::set_forward_axis(BoneAxis p_axis) {
```

File: scene/3d/look_at_modifier_3d.cpp [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+			return false;
+		}
+	} else if (Math::abs(p_prev[p_check_axis]) > origin_safe_margin && Math::abs(p_current[p_check_axis]) > origin_safe_margin) {
+		return false;
+	}
+
+	return signbit(p_prev[p_flipping_axis]) != signbit(p_current[p_flipping_axis]);
+}
+
+Vector3 LookAtModifier3D::get_basis_vector_from_bone_axis(Basis p_basis, LookAtModifier3D::BoneAxis p_axis) const {
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.cpp:559]:
```suggestion
Vector3 LookAtModifier3D::get_basis_vector_from_bone_axis(const Basis &p_basis, BoneAxis p_axis) const {
```

File: scene/3d/look_at_modifier_3d.cpp [hunk 5]
```diff
@@ -0,0 +1,781 @@
...
+			ret = p_basis.get_column(2);
+		} break;
+		case BONE_AXIS_MINUS_Z: {
+			ret = -p_basis.get_column(2);
+		} break;
+	}
+	return ret;
+}
+
+Vector3 LookAtModifier3D::get_vector_from_bone_axis(LookAtModifier3D::BoneAxis p_axis) const {
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.cpp:584]:
```suggestion
Vector3 LookAtModifier3D::get_vector_from_bone_axis(BoneAxis p_axis) const {
```

File: scene/3d/look_at_modifier_3d.cpp [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+	}
+	return ret;
+}
+
+float LookAtModifier3D::remap_damped(float p_from, float p_to, float p_damp_threshold, float p_value) const {
+	float sign = signbit(p_value) ? -1.0f : 1.0f;
+	float abs_value = Math::abs(p_value);
+
+	if (Math::is_equal_approx(p_damp_threshold, 1.0f) || Math::is_zero_approx(p_to)) {
+		return sign * CLAMP(abs_value, p_from, p_to); // Avoid zero division.
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.cpp:666]:
```suggestion
		return sign * CLAMP(abs_value, p_from, p_to); // Avoid division by zero.
```

File: scene/3d/look_at_modifier_3d.h [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+	float remain = 0;
+	float time_step = 1.0;
+
+	Vector3 get_basis_vector_from_bone_axis(Basis p_basis, BoneAxis p_axis) const;
+	Vector3 get_vector_from_bone_axis(BoneAxis p_axis) const;
+	Vector3 get_vector_from_axis(Vector3::Axis p_axis) const;
+	Vector3::Axis get_axis_from_bone_axis(BoneAxis p_axis) const;
+	Vector2 get_projection_vector(Vector3 p_vector, Vector3::Axis p_axis) const;
+	float remap_damped(float p_from, float p_to, float p_damp_threshold, float p_value) const;
+	double get_bspline_y(Vector2 p_from, Vector2 p_control, Vector2 p_to, double p_x) const;
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.h:111]:
```suggestion
	double get_bspline_y(const Vector2 &p_from, const Vector2 &p_control, Vector2 p_to, double p_x) const;
```

File: scene/3d/look_at_modifier_3d.h [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	bool is_intersecting_axis(Vector3 p_prev, Vector3 p_current, Vector3::Axis p_flipping_axis, Vector3::Axis p_check_axis, bool p_check_plane = false) const;
+
+	Transform3D look_at_with_axes(Transform3D p_rest);
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.h:114]:
```suggestion
	Transform3D look_at_with_axes(const Transform3D &p_rest);
```

File: scene/3d/look_at_modifier_3d.h [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+	void set_secondary_positive_limit_angle(float p_angle);
+	float get_secondary_positive_limit_angle() const;
+	void set_secondary_positive_damp_threshold(float p_power);
+	float get_secondary_positive_damp_threshold() const;
+	void set_secondary_negative_limit_angle(float p_angle);
+	float get_secondary_negative_limit_angle() const;
+	void set_secondary_negative_damp_threshold(float p_power);
+	float get_secondary_negative_damp_threshold() const;
+
+	float get_interpolation_remain() const;
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.h:191]:
```suggestion
	float get_interpolation_remaining() const;
```

File: scene/3d/look_at_modifier_3d.h [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	float secondary_negative_limit_angle = Math_PI;
+	float secondary_negative_damp_threshold = 1.0f;
+
+	bool is_within_limitations = false;
+
+	// For time-based interpolation.
+	Quaternion from_q;
+	Quaternion prev_q;
+
+	float remain = 0;
```
Comment (AThousandShips) [inline scene/3d/look_at_modifier_3d.h:102]:
```suggestion
	float remaining = 0;
```

**Comment by JekSun97** [comment]
@TokageItLab Thanks, I've been waiting for modifiers for skeletons for a long time!
I would like to know if there will be other modifiers in 4.4? If so, what will be their purpose?

**Comment by fire** [comment]
@JekSun97 Please review and contribute to the discussion here https://github.com/godotengine/godot-proposals/discussions/9885 (Needed Skeleton Modifiers)

**Comment by crei0** [comment]
I think there's a typo on the Documentation data `doc/classes/LookAtModifier3D.xml`, it's saying in some locations `SkeletonModifier3D` instead of the correct `LookAtModifier3D`

This is visible at the documentation page at [https://docs.godotengine.org/en/latest/classes/class_lookatmodifier3d.html](https://docs.godotengine.org/en/latest/classes/class_lookatmodifier3d.html)

Location 1:
`The SkeletonModifier3D rotates a bone to look a target.`
it should say
`The LookAtModifier3D rotates a bone to look a target.`

Location 2:
`This SkeletonModifier3D rotates a bone to look a target. [...]`
it should say
`This LookAtModifier3D rotates a bone to look a target. [...]`


And on `Create a new node` window on the tooltip and on the description section
image

**Pull request merged**